File: app-presentation/src/main/java/com/zxjdev/smile/presentation/authorization/register/RegisterPresenter.java
package com.zxjdev.smile.presentation.authorization.register;

import com.zxjdev.smile.domain.authorization.usecase.Register;
import com.zxjdev.smile.presentation.common.DefaultSubscriber;
import com.zxjdev.smile.presentation.common.util.ui.ErrorMessagePrinter;

import javax.inject.Inject;

public class RegisterPresenter implements RegisterContract.Presenter {

  private RegisterContract.View view;
  @Inject Register register;
  @Inject ErrorMessagePrinter errorMessagePrinter;

  @Inject
  RegisterPresenter() {

  }

  @Override
  public void handleRegister(String username, String password) {
    register.execute(new Register.RequestParams(username, password), new DefaultSubscriber<Void>(errorMessagePrinter) {
      @Override
      public void onComplete() {
        view.initUserComponent();
        view.navigateToMain();
      }
    });
  }

  @Override
  public void takeView(RegisterContract.View view) {
    this.view = view;
  }

  @Override
  public void dropView() {
    this.view = null;
  }
}

File: app-presentation/src/main/java/com/zxjdev/smile/presentation/authorization/register/RegisterPresenter.kt
package com.zxjdev.smile.presentation.authorization.register

import com.zxjdev.smile.domain.authorization.usecase.Register
import com.zxjdev.smile.presentation.common.util.ui.ErrorMessagePrinter
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableCompletableObserver
import javax.inject.Inject

class RegisterPresenter @Inject internal constructor() : RegisterContract.Presenter {

    private lateinit var view: RegisterContract.View
    @Inject lateinit var register: Register
    @Inject lateinit var errorMessagePrinter: ErrorMessagePrinter
    private val compositeDisposable = CompositeDisposable()

    override fun handleRegister(username: String, password: String) {
        compositeDisposable.add(register.execute(Register.RequestParams(username,
                password)).subscribeWith(object : DisposableCompletableObserver() {
            override fun onComplete() {
                view.initUserComponent()
                view.navigateToMain()
            }

            override fun onError(e: Throwable) {
                errorMessagePrinter.print(e.message)
            }
        }))
    }

    override fun takeView(view: RegisterContract.View) {
        this.view = view
    }

    override fun dropView() {
        compositeDisposable.dispose()
    }
}
